src/main/com/fulcrologic/fulcro/algorithms/normalized_state_helpers.cljc

```diff
@@ -0,0 +1,303 @@
+   (get-in state-map path nil))
+
+  ([state-map path not-found]
+   [map? vector? any? => any?]
+   (clojure.core/get-in state-map (tree-path->db-path state-map path) not-found)))
+
+
+(defn ui->props
+  "Obtain a tree of props for a UI instance from the current application state. Useful in mutations where you want
+  to denormalize an entity from the state database. `this` can often be obtained from the mutation `env` at the
+  `:component` key."
+  ([this]
+   (ui->props (comp/component->state-map this) (comp/react-type this) (comp/get-ident this)))
+  ([state-map component-class ident]
+   (fdn/db->tree (comp/get-query component-class state-map) (get-in state-map ident) state-map)))
+
+
+(defn- dissoc-in
+  "Dissociates an entry from a nested associative structure returning a new
+             nested structure. keys is a sequence of keys. Any empty maps that result
+             will not be present in the new structure."
+  [m [k & ks :as keys]]
+  (if ks
+    (if-let [nextmap (get m k)]
+      (let [newmap (dissoc-in nextmap ks)]
+        (if (seq newmap)
+          (assoc m k newmap)
+          (dissoc m k)))
+      m)
+    (dissoc m k)))
+
+
+(defn- normalized-paths
+  "Walks the tree in a depth first manner and returns the normalized possible paths"
+  [m]
+  (letfn [(paths* [ps ks m]
+            (reduce-kv
+              (fn [ps k v]
+                (if (map? v)
+                  (paths* ps (conj ks k) v)
+                  (conj ps (conj ks k))))
+              ps
+              m))]
+    (filter #(< (count %) 4)
+            (paths* () [] m))))
+
+
+(>defn remove-entity*
+  "Remove the given entity at the given ident. Also scans all tables and removes any to-one or to-many idents that are
+  found that match `ident` (removes dangling pointers to the removed entity).
+
+  The optional `cascade` parameter is a set of keywords that represent edges that should cause recursive deletes
+  (i.e. it indicates edge names that *own* something, indicating it is safe to remove those entities as well).
+
+  Returns the new state map with the entity(ies) removed."
+
+  ([state-map ident]
+   [map? eql/ident? => map?]
+   (remove-entity* state-map ident #{}))
+
+  ([state1 ident cascade]
+   [map? eql/ident? (s/coll-of keyword? :kind set?) => map?]
+   (let [tables (keep (fn [k]
+                        (let [candidate (get state1 k)]
+                          (when (and (map? candidate) (every? map? (vals candidate)))
+                            k))) (keys state1))
+
+         non-tables (keep (fn [k]
+                            (let [candidate (get state1 k)]
+                              (when (vector? candidate)
+                                [k])))
+                          (keys state1))
+
+         remove-idents-at-path (fn [state1 path]
+                                 (let [v (clojure.core/get-in state1 path)]
+                                   (cond
+                                     (int? v) state1
```

Comment:
Why isn't `int?` just the `else` case?

Reply:
Yeah, this does seem like a patchwork solution - working on it.